File: 3. Longest Substring Without Repeating Characters.py
# 3. Longest Substring Without Repeating Characters
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = {}
        i = 0
        ans = 0
        for j, c in enumerate(s):
            if c not in seen:
                ans = max(j - i + 1, ans)
            else:
                while i < j and s[i] != c:
                    del seen[s[i]]
                    i += 1
                i += 1
            seen[c] = j
        return ans

File: 3. Longest Substring Without Repeating Characters.cpp
// 3. Longest Substring Without Repeating Characters


#include <string>
#include <unordered_map>
using namespace std;

class Solution {
    public:
        int lengthOfLongestSubstring(string s) {
            unordered_map<char, int> map;
            int left = 0, ans = 0;
            int n = s.length();
            for (int right = 0; right < n; right++) {
                if (map.find(s[right]) != map.end() && map[s[right]] >= left) {
                    left = map[s[right]] + 1;
                }
                map[s[right]] = right;
                ans = max(ans, right - left + 1);
            }
            return ans;
        }
    };
    
    
    
    
    